Initial rate limiting impl based on the `leaky-bucket` and `concread` crates (#67)
memorysafety/river: user-manual/src/config/kdl.md

    * Arguments: `key="KEY" value="VALUE"`, where `KEY` is a valid HTTP header key, and `VALUE` is a valid HTTP header value
    * The given header will be added or replaced to `VALUE`

> ### `services.$NAME.rate-limiting`

Author: CC @branlwyd, added more detailed docs here as well!

This section contains the configuration for rate limiting rules.

Rate limiting rules are used to limit the total number of requests made by downstream clients,
based on various criteria.

Note that Rate limiting is on a **per service** basis, services do not share rate limiting
information.

This section is optional.

Example:

```
rate-limiting {
    timeout millis=100

    rule kind="source-ip" \
        max-buckets=4000 tokens-per-bucket=10 refill-qty=1 refill-rate-ms=10

    rule kind="uri" pattern="static/.*" \
        max-buckets=2000 tokens-per-bucket=20 refill-qty=5 refill-rate-ms=1
}
```

#### `services.$NAME.rate-limiting.timeout`

The `timeout` parameter is used to set the total timeout for acquiring all rate limiting tokens.

If acquiring applicable rate limiting tokens takes longer than this time, the request will not be
forwarded and will respond with a 429 error.

This parameter is mandatory if the `rate-limiting` section is present.

This is specified in the form:

`timeout millis=TIME`, where `TIME` is an unsigned integer

> **Implementation Note**: The rate limiting timeout is a temporary implementation detail to limit

Reviewer: Naive implementation question: why have a timeout, or a max debt, at all -- why can't the unavailability of a token immediately cause a 429?

Author: That's a good question! The hope was to support features like "delay" used [by nginx](https://blog.nginx.org/blog/rate-limiting-nginx).

However, until [leaky-bucket has support for limiting 'max debt'](https://github.com/udoprog/leaky-bucket/issues/34), it might be better to treat tokens as "now or never" like you proposed. I had missed it, but the `leaky-bucket` crate [does support this](https://docs.rs/leaky-bucket/latest/leaky_bucket/struct.RateLimiter.html#method.try_acquire), so I think I'll remove the timeout parameter, and only offer "request succeeds or fails immediately" for now!

instead of after a given timeout.

When this change occurs, the `timeout` parameter will be deprecated, and replaced with a `max-token-debt`
parameter instead.

#### `services.$NAME.rate-limiting.rule`

Rules are used to specify rate limiting parameters, and applicability of rules to a given request.

##### Leaky Buckets

Rate limiting in River uses a [Leaky Bucket] model for determining whether a request can be served
immediately, or if it should be delayed (or rejected). For a given rule, a "bucket" of "tokens"
is created, where one "token" is required for each request.

The bucket for a rule starts with a configurable `tokens-per-bucket` number. When a request arrives,
it attempts to take one token from the bucket. If one is available, it is served immediately. Otherwise,
the request waits in a first-in, first-out order for a token to become available.

The bucket is refilled at a configurable rate, specified by `refill-rate-ms`, and adds a configurable
number of tokens specified by `refill-qty`. The number of tokens in the bucket will never exceed the
initial `tokens-per-bucket` number.

> Once a refill occurs, requests may become ready if a token becomes available.

Reviewer: ```suggestion
Once the bucket is refilled, queued requests may take a token and begin processing.
```

(wordsmithing only, please check for accuracy)

Author: I think I'll decline this specific wording, because I want to make it clear that we don't have to wait until the bucket is FULLY refilled, e.g. if it has room for 10 tokens, and refills 2 tokens every millisecond, we'll serve two new requests every millisecond, we don't have to wait 5 milliseconds for the bucket to fully make it back to 10.

I'll take a look at rewording this line though!

Author: This went away when removing the concept of delay and timeouts

[Leaky Bucket]: https://en.wikipedia.org/wiki/Leaky_bucket

##### How many buckets?

Some rules require many buckets. For example, rules based on the source IP address will create a bucket

> for each unique IP address of downstream users.

Reviewer: ```suggestion
for each unique source IP address observed in a request.
```

(wordsmithing)

However, each of these buckets require space to contain the metadata, and to avoid unbounded growth,
we allow for a configurable `max-buckets` number, which serves to influence the total memory required
for storing buckets. This uses an [Adaptive Replacement Cache]

> to allow for concurrent access to these buckets, as well as the ability to automatically buckets that

Reviewer: ```suggestion
to allow for concurrent access to these buckets, as well as the ability to automatically evict buckets that
```

(missing word)

retain this information.

If `max-buckets` is set too low, then buckets will be "evicted" from the cache, meaning that subsequent
requests matching that bucket will require the creation of a new bucket (with a full set of tokens),
potentially defeating the objective of accurate rate limiting.

##### Gotta claim 'em all

When multiple rules apply to a single request, for example rules based on both source IP address,
and the URI path, then a request must claim ALL applicable tokens before proceeding. If a given IP
address is making it's first request, but to a URI that that has an empty bucket, it will immediately
obtain the IP address token, but be forced to wait until the URI token has been claimed

##### Kinds of Rules

Currently two kinds of rules are supported:

* `kind="source-ip"` - this tracks the IP address of the requestor. A unique bucket will be created for
  the IPv4 or IPv6 address of the requestor.
* `kind="uri" pattern="REGEX"` - This tracks the URI path of the request, such as `static/images/example.jpg`
    * If the request's URI path matches the provided `REGEX`, the full URI path will be assigned to a given
      bucket

>     * For example, if the regex `static/.*` was provided:

Reviewer: Q. What if there are multiple rules that match a given request URL? Is there an ordering/hierarchy of rules, or would the request match every rule with a matching regex?

Author: Right now, it is specified as "tokens must be acquired for all matching rules". So if you had two URI rules like:

* `static/.*`
* `.*\.jpg`

And requested `static/images/example.jpg`, you would match both rules, and need to get tokens for each rule.

I chose this behavior because it matches what NGINX does:

> You can include multiple limit_req directives in a single location. All limits that match a given request are applied, meaning the most restrictive one is used. For example, if more than one directive imposes a delay, the longest delay is used. Similarly, requests are rejected if that is the effect of any directive, even if other directives allow them through.

I'm open to other behaviors, but I don't have the context to know what other behavior  people could prefer here!

Author: I'm going to resolve this, but @branlwyd please do open an issue if there's another approach you'd prefer or would like to discuss in general!

Reviewer: Fair enough. I was thinking of a situation where you might want a low rate limit for most paths (e.g. `.*`), but a high rate limit for a few paths (e.g. `/static/.*`). In that case, you might want some way to only match the higher-limit rule, via hierarchy or priority ordering of rules. However, I _think_ such exclusions could also be implemented via a sufficiently-advanced-regex (e.g. in the previous example, the low-rate-limit rule would use a regex which matches everything but `/static/.*`), so this definitely is a nice-to-have.

        * `index.html` would not match this rule, and would not require obtaining a token
        * `static/images/example.jpg` would match this rule, and would require obtaining a token
        * `static/styles/example.css` would also match this rule, and would require obtaining a token

>         * Note that `static/images/example.jpg` and `static/styles/example.css` would each have a UNIQUE

Reviewer: I think (probably as an eventual feature) it would be very nice to rate-limit groups of endpoints together -- we do exactly this with our rate-limiter, and without reading docs I would guess a single path rule would have a single bucket used for all matching requests. I think we should be careful that we aren't painting ourselves into a corner here.

Author: This is a reasonable request, and I did go back and forth on the potential behavior! I could see it being useful to have a "single bucket rule" instead of a "multi bucket rule", like:

* one kind of rule that makes two buckets for `static/.*` for `static/images/example.jpg` and `static/styles/example.css`
* one kind of rule that makes exactly one bucket for `static/.*`, so `static/images/example.jpg` and `static/styles/example.css` take from the same bucket

I'll see if this is easy to specify in this PR, or open an issue to track future filter kinds.

Reviewer: Nice! Yep, this matches my expectations.
